Atomic upsert support for Postgresql

psibi opened the pull request:
I have created this PR for getting review.

Things that work now:

* Atomic upsert for postgresql work

Things that have to be done:

* Raise exception when there is more than one uniqueness constraint for an `EntityDef`.
* Fix other backend errors - Right now works only for postgresql

Question I have:

In what scenario will the field `uniqueFields` have more than one element in the list. Why it is modelled as a List data strcuture instead of a tuple like `(HaskellName, DBName)` ?

The `UniqueDef` type for reference:

``` haskell
data UniqueDef = UniqueDef
    { uniqueHaskell :: !HaskellName 
    , uniqueDBName  :: !DBName
    , uniqueFields  :: ![(HaskellName, DBName)] -- why list instead of single tuple ?
    , uniqueAttrs   :: ![Attr]
    }
    deriving (Show, Eq, Read, Ord)
```

gregwebs commented:
For `UniqueDef` that should be for a unique constraint that spans multiple fields

gregwebs commented:
Looking great! Once you write a test case everything else will fall into place.

gregwebs commented:
This will need to be re-based to the latest master now.

psibi commented:
@gregwebs Any idea on why the tests are failing: https://travis-ci.org/yesodweb/persistent/jobs/145239853 ?

The tests seem to be failing for `insert` sql query. I know that the TRAVIS postgresql version is 9.3 and we need a minimum of 9.5 to make this upsert sql feature work. But even then, the tests should fail with a different error message. Here it seems to get caught in a increasing series of `select` and `insert` queries.

gregwebs commented:
Looks like Travis only supports 9.4 right now.

I just pushed your branch with a merge to master which gets rid of sql debug logging on travis. I will look a the error there
https://travis-ci.org/yesodweb/persistent/builds/145356673

psibi commented:
There seems to be some workaround now for bringing 9.5: https://github.com/travis-ci/travis-ci/issues/4264#issuecomment-155404724

gregwebs commented:
I would rather stick with what Travis supports than start installing things. It avoids slowing down the build. The test on Travis will be good to prove that things still work as expected on < 9.5

psibi commented:
Okay. I see two options there:
* Wait till Postgresql 9.5 support comes in Travis. And then merge this PR.
* Merge the PR but pass `connUpsertSql = Nothing` in Postgresql configuration, so that the native upsert isn't used yet.

What do you recommend ? (Note that this PR has still some minor work left, so it's not ready for merge yet.)

gregwebs commented:
The persistent code needs to do a version check and only do a 9.5 upsert if postgresql is at version 9.5

gregwebs commented:
This is a great change! Should just need a squash and rebase at this point, which can be done from github if we just want 1 commit

psibi commented:
I have made some minor changes. I think this is good to merge now but I would like another review.

gregwebs commented on persistent-postgresql/Database/Persist/Postgresql.hs:
```diff
@@ -148,16 +149,38 @@ open' modConn cstr logFunc = do
 openSimpleConn :: (IsSqlBackend backend) => LogFunc -> PG.Connection -> IO backend
 openSimpleConn logFunc conn = do
     smap <- newIORef $ Map.empty
+    serverVersion <- getServerVersion conn
```
If `getServerVersion` should always work, then it is best to throw an exception here. And the result should be `IO (Either String Double)`, where the error case shows the version.

psibi replied:
I'm throwing an exception now. But I'm letting the type of `getServerVersion` be `IO (Maybe Double)`. I don't find any advantage of it being the `Either` data type as the exception is thrown when we encounter `Left` data.

gregwebs replied:
Where are you throwing an exception? It looks to me that if the version doesn't parse nothing will happen and a proper upsert will silently be unavailable. The point of using Either is so that the version string can be used in an error message.

gregwebs replied:
oh, sorry, I see it is on a later commit. In general I like to have base functions that return Either instead of throw an exception, but this will work.

gregwebs commented on persistent-postgresql/Database/Persist/Postgresql.hs:
```diff
@@ -199,6 +222,32 @@ insertSql' ent vals =
+                           , T.intercalate "," $ map (escape . fieldDB) $ entityFields ent
+                           , ") VALUES ("
+                           , T.intercalate "," $ map (const "?") (entityFields ent)
+                           , ") on conflict ("
```
Not sure if the rest of SQL is consistently using upper case.

psibi replied:
Thanks for the feedback. I made it upper case.

gregwebs commented on persistent/Database/Persist/Types/Base.hs:
```diff
@@ -218,8 +218,20 @@ toEmbedEntityDef ent = embDef
                         _ -> Nothing
                     }
 
+-- Type for storing the Uniqueness constraint in the Schema.
```
great docs, thanks!

gregwebs commented on persistent/Database/Persist/Class/PersistUnique.hs:
```diff
@@ -78,9 +78,6 @@ class (PersistUniqueRead backend, PersistStoreWrite backend) => PersistUniqueWri
            -- unique key)
            -> ReaderT backend m (Entity record)
            -- ^ the record in the database after the operation
-    upsert record updates = do
```
Does this need to be removed?

psibi replied:
No. I have added them back.

gregwebs commented:
Good to merge now? Do you want to do a rebase -i first? I see there is already a Changelog entry :)
I think it will be best to bump things to 2.6 since there is a change to the base persistent package affecting others.

psibi commented:
I have rebased the last 10 commits. But I think you can squash the remaining one from the github web interface itself. Free feel to do it. I have also changed the Changelog entry to point to 2.6. Haven't bumped the cabal version to 2.6. Let me know, if you want me to do that.

psibi commented:
@gregwebs Is any work still pending in this PR ? Can this be mergd.

gregwebs commented:
AWESOME! I will let this mellow on master for a little bit and then release by Monday.

gregwebs commented:
This is on hackage now

psibi commented:
:+1: Thanks!
